aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -0,0 +1,363 @@
+        // do not upload any files that have already been processed
+        .filter(Predicates.not(multiPartMap::containsKey))
+        .forEach(f -> {
+          UploadPartRequest.Builder requestBuilder = UploadPartRequest.builder()
+              .bucket(location.bucket())
+              .key(location.key())
+              .uploadId(multipartUploadId)
+              .partNumber(stagingFiles.indexOf(f) + 1)
+              .contentLength(f.length());
+
+          S3RequestUtil.configureEncryption(awsProperties, requestBuilder);
+
+          UploadPartRequest uploadRequest = requestBuilder.build();
+
+          CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
+              () -> {
+                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f));
+                return CompletedPart.builder().eTag(response.eTag()).partNumber(uploadRequest.partNumber()).build();
+              },
+              executorService
+          ).whenComplete((result, thrown) -> {
+            try {
+              Files.deleteIfExists(f.toPath());
+            } catch (IOException e) {
+              LOG.warn("Failed to delete staging file: {}", f, e);
+            }
+
+            if (thrown != null) {
+              LOG.error("Failed to upload part: {}", uploadRequest, thrown);
+              abortUpload();
+            }
+          });
+
+          multiPartMap.put(f, future);
+        });
+  }
+
+  private void completeMultiPartUpload() {
+    Preconditions.checkState(closed, "Complete upload called on open stream: " + location);
+
+    List<CompletedPart> completedParts =
+        multiPartMap.values()
+            .stream()
+            .map(CompletableFuture::join)
+            .sorted(Comparator.comparing(CompletedPart::partNumber))
+            .collect(Collectors.toList());
+
+    CompleteMultipartUploadRequest request = CompleteMultipartUploadRequest.builder()
+        .bucket(location.bucket()).key(location.key())
+        .uploadId(multipartUploadId)
+        .multipartUpload(CompletedMultipartUpload.builder().parts(completedParts).build()).build();
+
+    Tasks.foreach(request)
+        .noRetry()
+        .onFailure((r, thrown) -> {
+          LOG.error("Failed to complete multipart upload request: {}", r, thrown);
+          abortUpload();
+        })
+        .throwFailureWhenFinished()
+        .run(s3::completeMultipartUpload);
+  }
+
+  private void abortUpload() {
+    if (multipartUploadId != null) {
+      try {
+        s3.abortMultipartUpload(AbortMultipartUploadRequest.builder()
+            .bucket(location.bucket()).key(location.key()).uploadId(multipartUploadId).build());
+      } finally {
+        cleanUpStagingFiles();
+      }
+    }
+  }
+
+  private void cleanUpStagingFiles() {
+    Tasks.foreach(stagingFiles)
+        .suppressFailureWhenFinished()
+        .onFailure((file, thrown) -> LOG.warn("Failed to delete staging file: {}", file, thrown))
+        .run(File::delete);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of deleting a file on this line should be checked and failure should be logged. java.io.File.delete() will simply return false if it fails to delete the file, for example, when the file does not exist. It will not throw unless there is a security exception.